rfcs/convergence/native-element-props.md

```diff
@@ -0,0 +1,306 @@
+
+## Problem statement
+
+For most components, it makes sense to apply native props (and `ref`) to the root element. Currently the `root` slot always gets native element props passed to the component, and is not explicitly exposed via the API.
+
+The input components are a little unique because the actual `<input>` _cannot_ be the root element of the component, since we need various wrappers for styling (and `<input>` can't have children). However, users may expect that some or all native props passed to the root of the component would be applied to the `<input>` itself. This may also be a requirement for certain 3rd-party form libraries.
+
+For example, rendering a `<Checkbox/>` actually does this _(before the RFC is implemented)_:
+
+```tsx
+<slots.root>
+  <slots.checkmark {/* visual checkmark */} />
+  <slots.input {/* actual input */}>
+  {children /*label*/}
+</slots.root>
+```
+
+Or an `<Input/>` is roughly like this _(again pre-RFC)_:
+
+```tsx
+<slots.root>
+  <slots.bookendBefore />
+  <slots.inputWrapper {/* visually styled as the input */}>
+    <slots.insideStart />
+    <slots.input /> {/* actual input element */}
+    <slots.insideEnd />
+  </slots.inputWrapper>
+  <slots.bookendAfter />
+</slots.root>
+```
+
+So to specify a value, you'd have to use slot props:
+
+```tsx
+<Input input={{ id: 'foo', value: 'stuff' }} />
+```
+
+The question of where to apply the top-level `ref` has similar considerations and should likely follow the same pattern as native props.
+
+### Principles for any solution
+
+The chosen solution needs to adhere to the following principles (thanks @levithomason):
+
+- User needs to retain access to any part of DOM at any time (can't make a piece of DOM that's unable to receive props)
+- Consider what the user expects when using the component
+
+### For inputs: 3rd-party form validation libraries
+
+To the degree possible, we'd like inputs to work nicely with 3rd-party form libraries, which may have APIs that expect to be able to pass native props to the component and have them applied to the actual `<input>`. (needs more research)
+
+- [Formik](https://github.com/formium/formik)
+- [react-hook-form](https://github.com/react-hook-form/react-hook-form)
+  - Note, for this one they might need to make a wrapper for our component because the library expects to be able to _set_ `.value` on the input, which presents its own challenges. (Someone brought this up for v8 [here](https://github.com/microsoft/fluentui/issues/18126).)
+- [React Final Form](https://github.com/final-form/react-final-form)
+
+## Detailed Design or Proposal
+
+### Option A: allow setting a different "primary" slot; explicitly expose `root` _(formerly the proposed solution)_
+
+Allow setting a different "primary" slot (open to suggestions for the name) where top-level native props and `ref` are applied. This would default to `root` but could be customized (need to work with @bsunderhus to determine implementation approach).
```

Comment:
Clarification, `root` is always the root DOM element. It will usually be the same as "primary" (note that "primary" is NOT exposed to the user at all) except for special cases like inputs.